esphome/components/shelly_dimmer/shelly_dimmer.cpp

```diff
@@ -0,0 +1,476 @@
+    ESP_LOGW(TAG, "Failed to erase STM32 flash memory");
+    stm32_close(stm32);
+    return false;
+  }
+
+  const uint8_t *fw = STM_FIRMWARE;
+  uint32_t fw_len = sizeof(STM_FIRMWARE);
+
+  // Copy the STM32 firmware over in 256-byte chunks. Note that the firmware is stored
+  // in flash memory so all accesses need to be 4-byte aligned.
+  uint8_t buffer[256];
+  const uint8_t *p = fw;
+  uint32_t offset = 0;
+  uint32_t addr = stm32->dev->fl_start;
+  uint32_t end = addr + sizeof(STM_FIRMWARE);
+
+  while (addr < end && offset < fw_len) {
+    uint32_t left = end - addr;
+    uint32_t len = sizeof(buffer) > left ? left : sizeof(buffer);
+    len = (len > (fw_len - offset)) ? (fw_len - offset) : len;
+
+    if (len == 0) {
+      break;
+    }
+
+    memcpy(buffer, p, sizeof(buffer));
+    p += sizeof(buffer);
+
+    if (stm32_write_memory(stm32, addr, buffer, len) != STM32_ERR_OK) {
+      ESP_LOGW(TAG, "Failed to write to STM32 flash memory");
+      stm32_close(stm32);
+      return false;
+    }
+
+    addr += len;
+    offset += len;
+  }
+
+  stm32_close(stm32);
+  ESP_LOGI(TAG, "STM32 firmware upgrade successful");
+
+  return true;
+}
+
+uint16_t ShellyDimmer::convert_brightness_(float brightness) {
+  // Special case for zero as only zero means turn off completely.
+  if (brightness == 0.0) {
+    return 0;
+  }
+
+  uint16_t brightness_int = static_cast<uint16_t>(brightness * (this->max_brightness_ - this->min_brightness_));
+  brightness_int = brightness_int + this->min_brightness_;
+  return std::min(brightness_int, SHELLY_DIMMER_MAX_BRIGHTNESS);
+}
+
+void ShellyDimmer::send_brightness_(uint16_t brightness) {
+  uint8_t payload[SHELLY_DIMMER_PROTO_CMD_SWITCH_SIZE];
+  // Brightness (%) * 10.
+  payload[0] = brightness & 0xff, payload[1] = brightness >> 8,
+
+  this->send_command_(SHELLY_DIMMER_PROTO_CMD_SWITCH, payload, SHELLY_DIMMER_PROTO_CMD_SWITCH_SIZE);
+
+  this->brightness_ = brightness;
+}
+
+void ShellyDimmer::send_settings_() {
+  uint16_t fade_rate = std::min(uint16_t{100}, this->fade_rate_);
+
+  float brightness = 0.0;
+  if (this->state_ != nullptr) {
+    this->state_->current_values_as_brightness(&brightness);
+  }
+  uint16_t brightness_int = this->convert_brightness_(brightness);
+  ESP_LOGD(TAG, "Brightness update: %d (raw: %f)", brightness_int, brightness);
+
+  uint8_t payload[SHELLY_DIMMER_PROTO_CMD_SETTINGS_SIZE];
+  // Brightness (%) * 10.
+  payload[0] = brightness_int & 0xff;
+  payload[1] = brightness_int >> 8;
+  // Leading / trailing edge [0x01 = leading, 0x02 = trailing].
+  payload[2] = this->leading_edge_ ? 0x01 : 0x02;
+  payload[3] = 0x00;
+  // Fade rate.
+  payload[4] = fade_rate & 0xff;
+  payload[5] = fade_rate >> 8;
+  // Warmup brightness.
+  payload[6] = this->warmup_brightness_ & 0xff;
+  payload[7] = this->warmup_brightness_ >> 8;
+  // Warmup time.
+  payload[8] = this->warmup_time_ & 0xff;
+  payload[9] = this->warmup_time_ >> 8;
```

Comment:
```suggestion
  uint8_t payload[SHELLY_DIMMER_PROTO_CMD_SETTINGS_SIZE] = {
    // Brightness (%) * 10.
    static_cast<uint8_t>(brightness_int & 0xff),
    static_cast<uint8_t>(brightness_int >> 8),
    // Leading / trailing edge [0x01 = leading, 0x02 = trailing].
    this->leading_edge_ ? uint8_t{0x01} : uint8_t{0x02},
    0x00,
    // Fade rate.
    static_cast<uint8_t>(fade_rate & 0xff),
    static_cast<uint8_t>(fade_rate >> 8),
    // Warmup brightness.
    static_cast<uint8_t>(this->warmup_brightness_ & 0xff),
    static_cast<uint8_t>(this->warmup_brightness_ >> 8),
    // Warmup time.
    static_cast<uint8_t>(this->warmup_time_ & 0xff),
    static_cast<uint8_t>(this->warmup_time_ >> 8),
};
```
If `send_command_` took the payload by const, we could const the array as well (which may catch typos/errors).


An assert could also be used to enforce that all expected values are  there.
```
uint8_t payload[] = {
  ...
};
static_assert(sizeof(payload) == SHELLY_DIMMER_PROTO_CMD_SETTINGS_SIZE);
```

Reply:
Very nice, thank you!